src/app/rosetta/lib/block.ml

```diff
@@ -0,0 +1,579 @@
+     * *)
+    let plan : 'a Op.t list =
+      match t.kind with
+      | `Coinbase ->
+          [{Op.label= `Coinbase_inc; related_to= None}]
+      | `Fee_transfer ->
+          (* Detect if a fee transfer comes from a snark worker payment by
+         * failing to find the user command that caused it *)
+          if
+            List.find user_commands ~f:(fun (info : User_command_info.t) ->
+                Unsigned.UInt64.equal info.fee t.fee
+                && Unsigned.UInt64.equal info.token t.token
+                && [%eq: [`Pk of string]] info.receiver t.receiver )
+            |> Option.is_none
+          then
+            (* In this case we make a "synthetic" fee subtraction from block
+           * creator even though this is implicit in our blocks. We need this
+           * so the balance always sums to zero. *)
+            [ {Op.label= `Fee_payer_dec; related_to= None}
+            ; {Op.label= `Fee_receiver_inc; related_to= Some `Fee_payer_dec} ]
+          else
+            (* Otherwise we only have the receiver inc side *)
+            (* TODO: Relate fee_payer_dec with fee_receiver_inc across
+           * user_commands and internal commands *)
+            [{Op.label= `Fee_receiver_inc; related_to= Some `Fee_payer_dec}]
+    in
+    Op.build ~a_eq:[%eq: [`Coinbase_inc | `Fee_payer_dec | `Fee_receiver_inc]]
+      ~plan ~f:(fun ~related_operations ~operation_identifier op ->
+        (* All internal commands succeed if they're in blocks *)
+        let status = Operation_statuses.name `Success in
+        match op.label with
+        | `Coinbase_inc ->
+            { Operation.operation_identifier
+            ; related_operations
+            ; status
+            ; account=
+                Some (account_id block_creator_pk Amount_of.Token_id.default)
+            ; _type= Operation_types.name `Coinbase_inc
+            ; amount= Some (Amount_of.coda coinbase)
+            ; metadata= None }
+        | `Fee_payer_dec ->
+            { Operation.operation_identifier
+            ; related_operations
+            ; status
+            ; account= Some (account_id block_creator_pk t.token)
+            ; _type= Operation_types.name `Fee_payer_dec
+            ; amount= Some Amount_of.(negated @@ token t.token t.fee)
+            ; metadata= None }
+        | `Fee_receiver_inc ->
+            { Operation.operation_identifier
+            ; related_operations
+            ; status
+            ; account= Some (account_id t.receiver t.token)
+            ; _type= Operation_types.name `Fee_receiver_inc
+            ; amount= Some (Amount_of.token t.token t.fee)
+            ; metadata= None } )
+
+  let dummies =
+    [ { kind= `Coinbase
+      ; receiver= `Pk "Eve"
+      ; fee= Unsigned.UInt64.of_int 20_000_000_000
+      ; token= Unsigned.UInt64.of_int 1
+      ; hash= "COINBASE_1" } ]
+end
+
+module Block_info = struct
+  type t =
+    { block_identifier: Block_identifier.t
+    ; creator: [`Pk of string]
```

Comment:
rename `creator` to `block_reward_receiver`? or maybe add a separate field because these could be different pks

Reply:
When can these be different? Do we need to change the archive node schema to support this?

Here is what it is now:

```
CREATE TABLE blocks
( id                     serial PRIMARY KEY
, state_hash             text   NOT NULL UNIQUE
, parent_id              int                    REFERENCES blocks(id) ON DELETE SET NULL
, creator_id             int    NOT NULL        REFERENCES public_keys(id)
, snarked_ledger_hash_id int    NOT NULL        REFERENCES snarked_ledger_hashes(id)
, ledger_hash            text   NOT NULL
, height                 bigint NOT NULL
, timestamp              bigint NOT NULL
, coinbase_id            int                    REFERENCES internal_commands(id)
);
```

Reply:
Coinbase receiver is a daemon flag and can be different from the block producer key (creator). 
In the archive node schema we store the receiver pk for coinbase and fee transfer that would be of the coinbase receiver. However, in here we are using block_creator_pk for `Coinbase_inc`. So maybe just use the receiver pk from the coinbase transaction. 
But maybe it is good to add another field for the coinbase receiver and differentiate these two at the top-level to avoid any confusion?